src/app/codegen-data-model/CodegenDataModel_Write.cpp

```diff
@@ -321,45 +332,68 @@ CHIP_ERROR CodegenDataModel::WriteAttribute(const InteractionModel::WriteAttribu
         }
     }
 
     AttributeAccessInterface * aai       = GetAttributeAccessOverride(request.path.mEndpointId, request.path.mClusterId);
     std::optional<CHIP_ERROR> aai_result = TryWriteViaAccessInterface(request.path, aai, decoder);
-
     if (aai_result.has_value())
     {
+        if (*aai_result == CHIP_NO_ERROR)
+        {
+            // TODO: change callbacks should likely be routed through the context `MarkDirty` only
+            //       however for now this is called directly because ember code does this call
+            //       inside emberAfWriteAttribute.
+            MatterReportingAttributeChangeCallback(request.path);
+            CurrentContext().dataModelChangeListener->MarkDirty(request.path);
+        }
         return *aai_result;
     }
 
-    ReturnErrorCodeIf(aai_result.has_value(), *aai_result);
+    MutableByteSpan dataBuffer = gEmberAttributeIOBufferSpan;
+    ReturnErrorOnFailure(DecodeValueIntoEmberBuffer(decoder, *attributeMetadata, dataBuffer));
 
-    ReturnErrorOnFailure(DecodeValueIntoEmberBuffer(decoder, *attributeMetadata, gEmberAttributeIOBufferSpan));
+    Protocols::InteractionModel::Status status;
 
-    EmberAfAttributeSearchRecord record;
-    record.endpoint    = request.path.mEndpointId;
-    record.clusterId   = request.path.mClusterId;
-    record.attributeId = request.path.mAttributeId;
+    if (request.operationFlags.Has(InteractionModel::OperationFlags::kInternal))
+    {
+        // Internal requests use the non-External interface that has less enforcement
+        // than the external version (e.g. does not check/enforce writable settings, does not
+        // validate atribute types) - see attribute-table.h documentation for details.
+        status = emberAfWriteAttribute(request.path.mEndpointId, request.path.mClusterId, request.path.mAttributeId,
+                                       gEmberAttributeIOBufferSpan.data(), (*attributeMetadata)->attributeType);
+    }
+    else
+    {
+        if (dataBuffer.size() > (*attributeMetadata)->size)
+        {
+            ChipLogDetail(Zcl, "Data to write exceedes the attribute size claimed.");
+            return CHIP_IM_GLOBAL_STATUS(InvalidValue);
+        }
 
-    Protocols::InteractionModel::Status status = emAfReadOrWriteAttribute(
-        &record, attributeMetadata, gEmberAttributeIOBufferSpan.data(), static_cast<uint16_t>(gEmberAttributeIOBufferSpan.size()),
-        /* write = */ true);
+        status = emAfWriteAttributeExternal(request.path.mEndpointId, request.path.mClusterId, request.path.mAttributeId,
+                                            gEmberAttributeIOBufferSpan.data(), (*attributeMetadata)->attributeType);
+    }
 
     if (status != Protocols::InteractionModel::Status::Success)
     {
         return CHIP_ERROR_IM_GLOBAL_STATUS_VALUE(status);
     }
 
+    // TODO: this may need more refinement:
+    //       - should internal requests be able to decide if something is marked dirty or not?
```

Comment:
The answer is 100% yes, and this is critical to achieve the Q quality.

See https://github.com/project-chip/connectedhomeip/pull/31162

Reply:
yes, however we currently do not have this capability in ember (we do have non-dirty sets in accessors, so we have to make that work, however reality is we should just support the Q quality).

The point of this todo was that this will need more work, however in the current infrastructure is it not possible, until we integrate this code and can change both sides of the code (accessors and this).